File: app/controllers/iop/transfer/impl/ExternalExaminationController.java
/*
 * Copyright (c) 2018 The members of the EXAM Consortium (https://confluence.csc.fi/display/EXAM/Konsortio-organisaatio)
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package controllers.iop.transfer.impl;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import controllers.ExaminationController;
import controllers.iop.transfer.api.ExternalAttachmentLoader;
import impl.AutoEvaluationHandler;
import impl.EmailComposer;
import io.ebean.DB;
import io.ebean.text.PathProperties;
import io.vavr.Tuple;
import io.vavr.Tuple2;
import io.vavr.control.Either;
import java.io.IOException;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import java.util.stream.StreamSupport;
import javax.inject.Inject;
import models.Exam;
import models.ExamEnrolment;
import models.User;
import models.json.ExternalExam;
import models.questions.ClozeTestAnswer;
import models.questions.EssayAnswer;
import models.questions.Question;
import models.sections.ExamSectionQuestion;
import org.apache.pekko.actor.ActorSystem;
import org.joda.time.DateTime;
import play.Environment;
import play.libs.concurrent.ClassLoaderExecutionContext;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.With;
import repository.ExaminationRepository;
import sanitizers.Attrs;
import sanitizers.EssayAnswerSanitizer;
import security.Authenticated;
import system.interceptors.SensitiveDataPolicy;
import util.config.ByodConfigHandler;
import util.datetime.DateTimeHandler;

@SensitiveDataPolicy(sensitiveFieldNames = { "score", "defaultScore", "correctOption", "configKey" })
@Restrict({ @Group("STUDENT") })
public class ExternalExaminationController extends ExaminationController {

    @Inject
    public ExternalExaminationController(
        EmailComposer emailComposer,
        ActorSystem actor,
        ExaminationRepository examinationRepository,
        AutoEvaluationHandler autoEvaluationHandler,
        Environment environment,
        ClassLoaderExecutionContext httpExecutionContext,
        ExternalAttachmentLoader externalAttachmentLoader,
        ByodConfigHandler byodConfigHandler,
        DateTimeHandler dateTimeHandler
    ) {
        super(
            emailComposer,
            examinationRepository,
            actor,
            autoEvaluationHandler,
            environment,
            httpExecutionContext,
            externalAttachmentLoader,
            byodConfigHandler,
            dateTimeHandler
        );
    }

    @Authenticated
    @Override
    public CompletionStage<Result> startExam(String hash, Http.Request request) throws IOException {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        Optional<ExternalExam> optional = getExternalExam(hash, user);
        if (optional.isEmpty()) {
            return wrapAsPromise(forbidden());
        }
        ExternalExam externalExam = optional.get();
        Optional<ExamEnrolment> optionalEnrolment = getEnrolment(user, externalExam);
        if (optionalEnrolment.isEmpty()) {
            return wrapAsPromise(forbidden());
        }
        ExamEnrolment enrolment = optionalEnrolment.get();
        Exam newExam = externalExam.deserialize();
        return getEnrolmentError(enrolment, request)
            .thenApplyAsync(error -> {
                if (error.isPresent()) {
                    return error.get();
                }
                if (newExam.getState().equals(Exam.State.PUBLISHED)) {
                    newExam.setState(Exam.State.STUDENT_STARTED);
                    try {
                        externalExam.serialize(newExam);
                    } catch (IOException e) {
                        return internalServerError();
                    }
                    DateTime now = dateTimeHandler.adjustDST(
                        DateTime.now(),
                        enrolment.getReservation().getMachine().getRoom()
                    );
                    externalExam.setStarted(now);
                    externalExam.update();
                }
                newExam.setCloned(false);
                newExam.setExternal(true);
                newExam.setDerivedMaxScores();
                processClozeTestQuestions(newExam);
                return ok(newExam, getPath(false));
            });
    }

    private void processClozeTestQuestions(Exam exam) {
        Set<Question> questionsToHide = new HashSet<>();
        exam
            .getExamSections()
            .stream()
            .flatMap(es -> es.getSectionQuestions().stream())
            .filter(esq -> esq.getQuestion().getType() == Question.Type.ClozeTestQuestion)
            .forEach(esq -> {
                ClozeTestAnswer answer = esq.getClozeTestAnswer();
                if (answer == null) {
                    answer = new ClozeTestAnswer();
                }
                answer.setQuestion(esq);
                esq.setClozeTestAnswer(answer);
                questionsToHide.add(esq.getQuestion());
            });
        questionsToHide.forEach(q -> q.setQuestion(null));
    }

    @Authenticated
    @Override
    public CompletionStage<Result> turnExam(String hash, Http.Request request) {
        return CompletableFuture.completedFuture(
            terminateExam(hash, Exam.State.REVIEW, request.attrs().get(Attrs.AUTHENTICATED_USER))
        );
    }

    @Authenticated
    @Override
    public CompletionStage<Result> abortExam(String hash, Http.Request request) {
        return CompletableFuture.completedFuture(
            terminateExam(hash, Exam.State.ABORTED, request.attrs().get(Attrs.AUTHENTICATED_USER))
        );
    }

    @Authenticated
    @Override
    public CompletionStage<Result> answerMultiChoice(String hash, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        return getEnrolmentError(hash, user, request)
            .thenApplyAsync(err ->
                err.orElseGet(() -> {
                    Optional<ExternalExam> optional = getExternalExam(hash, user);
                    if (optional.isEmpty()) {
                        return forbidden();
                    }
                    ExternalExam ee = optional.get();
                    ArrayNode node = (ArrayNode) request.body().asJson().get("oids");
                    List<Long> optionIds = StreamSupport
                        .stream(node.spliterator(), false)
                        .map(JsonNode::asLong)
                        .toList();
                    return findSectionQuestion(ee, qid)
                        .map(t -> processOptions(optionIds, t._2, ee, t._1))
                        .getOrElseGet(Function.identity());
                })
            );
    }

    @Authenticated
    @With(EssayAnswerSanitizer.class)
    @Override
    public CompletionStage<Result> answerEssay(String hash, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        return getEnrolmentError(hash, user, request)
            .thenApplyAsync(err ->
                err.orElseGet(() -> {
                    Optional<ExternalExam> optional = getExternalExam(hash, user);
                    if (optional.isEmpty()) {
                        return forbidden();
                    }
                    ExternalExam ee = optional.get();
                    String essayAnswer = request.attrs().getOptional(Attrs.ESSAY_ANSWER).orElse(null);
                    Optional<Long> objectVersion = request.attrs().getOptional(Attrs.OBJECT_VERSION);
                    Optional<ExamSectionQuestion> optionalQuestion;
                    Exam content;
                    try {
                        content = ee.deserialize();
                        optionalQuestion = findQuestion(qid, content);
                    } catch (IOException e) {
                        return internalServerError();
                    }
                    if (optionalQuestion.isEmpty()) {
                        return forbidden();
                    }
                    ExamSectionQuestion question = optionalQuestion.get();
                    EssayAnswer answer = question.getEssayAnswer();
                    if (answer == null) {
                        answer = new EssayAnswer();
                    } else if (objectVersion.isPresent()) {
                        if (answer.getObjectVersion() > objectVersion.get()) {
                            // Optimistic locking problem
                            return forbidden("sitnet_error_data_has_changed");
                        }
                        answer.setObjectVersion(objectVersion.get() + 1);
                    }
                    answer.setAnswer(essayAnswer);
                    question.setEssayAnswer(answer);
                    try {
                        ee.serialize(content);
                    } catch (IOException e) {
                        return internalServerError();
                    }
                    return ok(answer);
                })
            );
    }

    private Either<Result, Tuple2<Exam, ExamSectionQuestion>> findSectionQuestion(ExternalExam ee, Long qid) {
        Optional<ExamSectionQuestion> optionalQuestion;
        Exam content;
        try {
            content = ee.deserialize();
            optionalQuestion = findQuestion(qid, content);
        } catch (IOException e) {
            return Either.left(internalServerError());
        }
        return optionalQuestion
            .<Either<Result, Tuple2<Exam, ExamSectionQuestion>>>map(examSectionQuestion ->
                Either.right(Tuple.of(content, examSectionQuestion))
            )
            .orElseGet(() -> Either.left(forbidden()));
    }

    @Authenticated
    @With(EssayAnswerSanitizer.class)
    @Override
    public CompletionStage<Result> answerClozeTest(String hash, Long qid, Http.Request request) {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        return getEnrolmentError(hash, user, request)
            .thenApplyAsync(err ->
                err.orElseGet(() -> {
                    Optional<ExternalExam> optional = getExternalExam(hash, user);
                    if (optional.isEmpty()) {
                        return forbidden();
                    }
                    ExternalExam ee = optional.get();
                    return findSectionQuestion(ee, qid)
                        .map(t -> {
                            ExamSectionQuestion esq = t._2;
                            ClozeTestAnswer answer = esq.getClozeTestAnswer();
                            if (answer == null) {
                                answer = new ClozeTestAnswer();
                                esq.setClozeTestAnswer(answer);
                            } else {
                                long objectVersion = request.attrs().get(Attrs.OBJECT_VERSION);
                                if (answer.getObjectVersion() > objectVersion) {
                                    // Optimistic locking problem
                                    return forbidden("sitnet_error_data_has_changed");
                                }
                                answer.setObjectVersion(objectVersion + 1);
                            }
                            answer.setAnswer(request.attrs().getOptional(Attrs.ESSAY_ANSWER).orElse(null));
                            try {
                                ee.serialize(t._1);
                            } catch (IOException e) {
                                return internalServerError();
                            }
                            return ok(answer, PathProperties.parse("(id, objectVersion, answer)"));
                        })
                        .getOrElseGet(Function.identity());
                })
            );
    }

    private Optional<ExamSectionQuestion> findQuestion(Long qid, Exam content) {
        return content
            .getExamSections()
            .stream()
            .flatMap(es -> es.getSectionQuestions().stream())
            .filter(esq -> esq.getId().equals(qid))
            .findFirst();
    }

    private Result processOptions(List<Long> oids, ExamSectionQuestion esq, ExternalExam ee, Exam content) {
        esq.getOptions().forEach(o -> o.setAnswered(oids.contains(o.getId())));
        try {
            ee.serialize(content);
        } catch (IOException e) {
            return internalServerError();
        }
        return ok();
    }

    private Optional<ExternalExam> getExternalExam(String hash, User user) {
        return DB.find(ExternalExam.class).where().eq("hash", hash).eq("creator", user).forUpdate().findOneOrEmpty();
    }

    private Optional<ExamEnrolment> getEnrolment(User user, ExternalExam prototype) {
        DateTime now = dateTimeHandler.adjustDST(DateTime.now());
        ExamEnrolment enrolment = DB
            .find(ExamEnrolment.class)
            .fetch("reservation")
            .fetch("reservation.machine")
            .fetch("reservation.machine.room")
            .where()
            .eq("user.id", user.getId())
            .eq("externalExam.hash", prototype.getHash())
            .le("reservation.startAt", now.toDate())
            .gt("reservation.endAt", now.toDate())
            .eq("reservation.externalUserRef", user.getEppn())
            .findOne();
        return Optional.ofNullable(enrolment);
    }

    private CompletionStage<Optional<Result>> getEnrolmentError(String hash, User user, Http.Request request) {
        ExamEnrolment enrolment = DB
            .find(ExamEnrolment.class)
            .where()
            .eq("externalExam.hash", hash)
            .eq("externalExam.creator", user)
            .jsonEqualTo("externalExam.content", "state", Exam.State.STUDENT_STARTED.toString())
            .findOne();
        return getEnrolmentError(enrolment, request);
    }

    private Result terminateExam(String hash, Exam.State newState, User user) {
        ExternalExam ee = DB.find(ExternalExam.class).where().eq("hash", hash).eq("creator", user).findOne();
        if (ee == null) {
            return forbidden();
        }
        Optional<ExamEnrolment> optionalEnrolment = getEnrolment(user, ee);
        if (optionalEnrolment.isEmpty()) {
            return forbidden();
        }
        ExamEnrolment enrolment = optionalEnrolment.get();
        DateTime now = dateTimeHandler.adjustDST(DateTime.now(), enrolment.getReservation().getMachine().getRoom());
        ee.setFinished(now);
        try {
            Exam content = ee.deserialize();
            if (content.getState().equals(Exam.State.STUDENT_STARTED)) {
                content.setState(newState);
            }
            ee.serialize(content);
        } catch (IOException e) {
            return internalServerError();
        }
        return ok();
    }
}

File: app/controllers/iop/transfer/impl/ExternalExaminationController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl

import controllers.examination.{EnrolmentValidator, ExaminationController}
import io.ebean.DB
import io.ebean.text.PathProperties
import miscellaneous.config.ByodConfigHandler
import miscellaneous.datetime.DateTimeHandler
import miscellaneous.scala.{DbApiHelper, JavaApiHelper}
import models.enrolment.ExamEnrolment
import models.exam.Exam
import models.iop.ExternalExam
import models.questions.{ClozeTestAnswer, EssayAnswer, Question}
import models.sections.ExamSectionQuestion
import models.user.{Role, User}
import org.joda.time.DateTime
import play.api.{Environment, Logging}
import play.api.libs.json.JsValue
import play.api.mvc.*
import security.scala.Auth
import security.scala.Auth.{AuthenticatedAction, authorized}
import system.AuditedAction
import system.interceptors.scala.{SecureController, SensitiveDataFilter}
import validation.scala.answer.{ClozeTestAnswerValidator, EssayAnswerValidator}

import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters.*
import scala.util.Try

class ExternalExaminationController @Inject() (
    dateTimeHandler: DateTimeHandler,
    byodConfigHandler: ByodConfigHandler,
    val environment: Environment,
    val authenticated: AuthenticatedAction,
    val audited: AuditedAction,
    val sensitiveDataFilter: SensitiveDataFilter,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends SecureController
    with EnrolmentValidator
    with DbApiHelper
    with JavaApiHelper
    with Logging:

  override protected def sensitiveFields: Set[String] =
    Set("score", "defaultScore", "correctOption", "configKey")

  def startExam(hash: String): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.STUDENT))).async { request =>
      val user = request.attrs(Auth.ATTR_USER)

      getExternalExam(hash, user) match
        case None => Future.successful(Forbidden)
        case Some(externalExam) =>
          getEnrolment(user, externalExam) match
            case None => Future.successful(Forbidden)
            case Some(enrolment) =>
              Try(externalExam.deserialize()) match
                case scala.util.Failure(_) => Future.successful(InternalServerError)
                case scala.util.Success(newExam) =>
                  validateBasicEnrolment(enrolment, request).map {
                    case Some(error) => error
                    case None =>
                      if newExam.getState == Exam.State.PUBLISHED then
                        newExam.setState(Exam.State.STUDENT_STARTED)
                        Try(externalExam.serialize(newExam)) match
                          case scala.util.Failure(_) => InternalServerError
                          case scala.util.Success(_) =>
                            val now = dateTimeHandler.adjustDST(
                              DateTime.now(),
                              enrolment.getReservation.getMachine.getRoom
                            )
                            externalExam.setStarted(now)
                            externalExam.update()
                            processExamForResponse(newExam)
                      else processExamForResponse(newExam)
                  }
    }

  private def processExamForResponse(exam: Exam) =
    exam.setCloned(false)
    exam.setExternal(true)
    exam.setDerivedMaxScores()
    processClozeTestQuestions(exam)
    Ok(exam.asJson(ExaminationController.getPath(false)))

  private def processClozeTestQuestions(exam: Exam): Unit =
    val clozeTestQuestions = exam.getExamSections.asScala
      .flatMap(_.getSectionQuestions.asScala)
      .filter(_.getQuestion.getType == Question.Type.ClozeTestQuestion)
      .toList

    val questionsToHide = clozeTestQuestions.map { esq =>
      val answer = Option(esq.getClozeTestAnswer).getOrElse(new ClozeTestAnswer())
      answer.setQuestion(esq)
      esq.setClozeTestAnswer(answer)
      esq.getQuestion
    }.toSet
    questionsToHide.foreach(_.setQuestion(null))

  def turnExam(hash: String): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.STUDENT))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      terminateExam(hash, Exam.State.REVIEW, user)
    }

  def abortExam(hash: String): Action[AnyContent] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.STUDENT))) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      terminateExam(hash, Exam.State.ABORTED, user)
    }

  def answerMultiChoice(hash: String, qid: Long): Action[JsValue] =
    audited.andThen(authenticated).andThen(authorized(Seq(Role.Name.STUDENT))).async(parse.json) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      getEnrolmentError(hash, user, request).map {
        case Some(error) => error
        case None =>
          getExternalExam(hash, user) match
            case None => Forbidden
            case Some(ee) =>
              val optionIds = (request.body \ "oids").asOpt[Seq[Long]].getOrElse(Seq.empty).toList
              findSectionQuestion(ee, qid) match
                case Left(error)           => error
                case Right((content, esq)) => processOptions(optionIds, esq, ee, content)
      }
    }

  def answerEssay(hash: String, qid: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.STUDENT)))
    .andThen(EssayAnswerValidator.filter)
    .async(parse.json) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      getEnrolmentError(hash, user, request).flatMap {
        case Some(error) => Future.successful(error)
        case None =>
          val answerDTO = request.attrs(EssayAnswerValidator.ESSAY_ANSWER_KEY)
          getExternalExam(hash, user) match
            case None => Future.successful(Forbidden)
            case Some(ee) =>
              Try(ee.deserialize()) match
                case scala.util.Failure(_) => Future.successful(InternalServerError)
                case scala.util.Success(content) =>
                  findQuestion(qid, content) match
                    case None => Future.successful(Forbidden)
                    case Some(question) =>
                      val answer = Option(question.getEssayAnswer).getOrElse(new EssayAnswer())
                      // Handle optimistic locking
                      val hasVersionConflict = answerDTO.objectVersion match
                        case Some(ov) if answer.getObjectVersion > ov => true
                        case Some(ov) =>
                          answer.setObjectVersion(ov + 1)
                          false
                        case None => false
                      if hasVersionConflict then Future.successful(Forbidden("i18n_error_data_has_changed"))
                      else
                        answer.setAnswer(answerDTO.answer)
                        question.setEssayAnswer(answer)

                        Try(ee.serialize(content)) match
                          case scala.util.Failure(_) => Future.successful(InternalServerError)
                          case scala.util.Success(_) => Future.successful(Ok(answer.asJson))
      }
    }

  def answerClozeTest(hash: String, qid: Long): Action[JsValue] = audited
    .andThen(authenticated)
    .andThen(authorized(Seq(Role.Name.STUDENT)))
    .andThen(ClozeTestAnswerValidator.filter)
    .async(parse.json) { request =>
      val user = request.attrs(Auth.ATTR_USER)
      getEnrolmentError(hash, user, request).flatMap {
        case Some(error) => Future.successful(error)
        case None =>
          val answerDTO     = request.attrs(ClozeTestAnswerValidator.CLOZE_TEST_ANSWER_KEY)
          val objectVersion = answerDTO.objectVersion.getOrElse(0L)
          getExternalExam(hash, user) match
            case None => Future.successful(Forbidden)
            case Some(ee) =>
              findSectionQuestion(ee, qid) match
                case Left(error) => Future.successful(error)
                case Right((content, esq)) =>
                  val answer = Option(esq.getClozeTestAnswer).getOrElse {
                    val newAnswer = new ClozeTestAnswer()
                    esq.setClozeTestAnswer(newAnswer)
                    newAnswer
                  }

                  // Handle optimistic locking
                  if answer.getObjectVersion > objectVersion then
                    Future.successful(Forbidden("i18n_error_data_has_changed"))
                  else
                    answer.setObjectVersion(objectVersion + 1)
                    answer.setAnswer(answerDTO.answer)

                    Try(ee.serialize(content)) match
                      case scala.util.Failure(_) => Future.successful(InternalServerError)
                      case scala.util.Success(_) =>
                        Future.successful(Ok(answer.asJson(PathProperties.parse("(id, objectVersion, answer)"))))
      }
    }

  private def findSectionQuestion(ee: ExternalExam, qid: Long): Either[Result, (Exam, ExamSectionQuestion)] =
    Try(ee.deserialize()) match
      case scala.util.Failure(_) => Left(InternalServerError)
      case scala.util.Success(content) =>
        findQuestion(qid, content) match
          case None      => Left(Forbidden)
          case Some(esq) => Right((content, esq))

  private def findQuestion(qid: Long, content: Exam): Option[ExamSectionQuestion] =
    content.getExamSections.asScala
      .flatMap(_.getSectionQuestions.asScala)
      .find(_.getId == qid)

  private def processOptions(oids: List[Long], esq: ExamSectionQuestion, ee: ExternalExam, content: Exam): Result =
    esq.getOptions.asScala.foreach(o => o.setAnswered(oids.contains(o.getId)))
    Try(ee.serialize(content)) match
      case scala.util.Failure(_) => InternalServerError
      case scala.util.Success(_) => Ok

  private def getExternalExam(hash: String, user: User): Option[ExternalExam] =
    DB.find(classOf[ExternalExam])
      .where()
      .eq("hash", hash)
      .eq("creator", user)
      .forUpdate()
      .find

  private def getEnrolment(user: User, prototype: ExternalExam): Option[ExamEnrolment] =
    val now = dateTimeHandler.adjustDST(DateTime.now())
    DB.find(classOf[ExamEnrolment])
      .fetch("reservation")
      .fetch("reservation.machine")
      .fetch("reservation.machine.room")
      .where()
      .eq("user.id", user.getId)
      .eq("externalExam.hash", prototype.getHash)
      .le("reservation.startAt", now.toDate)
      .gt("reservation.endAt", now.toDate)
      .eq("reservation.externalUserRef", user.getEppn)
      .find

  private def getEnrolmentError(hash: String, user: User, request: Request[JsValue]) =
    DB.find(classOf[ExamEnrolment])
      .where()
      .eq("externalExam.hash", hash)
      .eq("externalExam.creator", user)
      .jsonEqualTo("externalExam.content", "state", Exam.State.STUDENT_STARTED.toString)
      .find match
      case None    => Future.successful(None)
      case Some(e) => validateBasicEnrolment(e, request)

  private def terminateExam(hash: String, newState: Exam.State, user: User): Result =
    DB.find(classOf[ExternalExam])
      .where()
      .eq("hash", hash)
      .eq("creator", user)
      .find match
      case None => Forbidden
      case Some(ee) =>
        getEnrolment(user, ee) match
          case None => Forbidden
          case Some(enrolment) =>
            val now = dateTimeHandler.adjustDST(DateTime.now(), enrolment.getReservation.getMachine.getRoom)
            ee.setFinished(now)
            Try(ee.deserialize()) match
              case scala.util.Failure(_) => InternalServerError
              case scala.util.Success(content) =>
                if content.getState == Exam.State.STUDENT_STARTED then content.setState(newState)
                Try(ee.serialize(content)) match
                  case scala.util.Failure(_) => InternalServerError
                  case scala.util.Success(_) => Ok
